pkg/controller/servicebindingrequest/annotations/resource.go

```diff
@@ -0,0 +1,208 @@
+// the resulting slice.
+func getInputPathFields(bindingInfo *BindingInfo, inputPathPrefix *string) []string {
+	inputPathFields := []string{}
+	if bindingInfo.ResourceReferencePath != bindingInfo.SourcePath {
+		inputPathFields = append(inputPathFields, bindingInfo.SourcePath)
+	}
+	if inputPathPrefix != nil && len(*inputPathPrefix) > 0 {
+		inputPathFields = append([]string{*inputPathPrefix}, inputPathFields...)
+	}
+	return inputPathFields
+}
+
+// Handle returns the value for an external resource strategy.
+func (h *ResourceHandler) Handle() (Result, error) {
+	ns := h.resource.GetNamespace()
+	resource, err := h.
+		client.
+		Resource(h.relatedGroupVersionResource).
+		Namespace(ns).
+		Get(h.relatedResourceName, metav1.GetOptions{})
+	if err != nil {
+		return Result{}, fmt.Errorf("error handling annotation: %w", err)
+	}
+
+	inputPathFields := getInputPathFields(h.bindingInfo, h.inputPathRoot)
+	val, ok, err := unstructured.NestedFieldCopy(resource.Object, inputPathFields...)
+	if !ok {
+		return Result{}, InvalidArgumentErr(strings.Join(inputPathFields, ", "))
+	}
+	if err != nil {
+		return Result{}, err
+	}
+
+	if mapVal, ok := val.(map[string]interface{}); ok {
+		tmpVal := make(map[string]interface{})
+		for k, v := range mapVal {
+			decodedVal, err := h.stringValue(v)
+			if err != nil {
+				return Result{}, err
+			}
+			tmpVal[k] = decodedVal
+		}
+		val = tmpVal
+	} else {
+		val, err = h.stringValue(val)
+		if err != nil {
+			return Result{}, err
+		}
+	}
+
+	typ, err := discoverBindingType(h.bindingInfo.Value)
+	if err != nil {
+		return Result{}, err
+	}
+
+	// get resource's kind.
+	gvk, err := h.restMapper.KindFor(h.relatedGroupVersionResource)
+	if err != nil {
+		return Result{}, err
+	}
+
+	// prefix the output path with the kind of the resource.
+	outputPath := strings.Join([]string{
+		strings.ToLower(gvk.Kind),
+		h.bindingInfo.SourcePath,
+	}, ".")
+
+	return Result{
+		Object: nested.ComposeValue(val, nested.NewPath(outputPath)),
+		Type:   typ,
+		Path:   outputPath,
+	}, nil
+}
+
+// stringValue asserts the given value 'v' and returns its string value.
+func stringValue(v interface{}) (string, error) {
+	s, ok := v.(string)
+	if !ok {
+		return "", fmt.Errorf("value is not a string")
+	}
+	return s, nil
+}
+
+// NewSecretHandler constructs a SecretHandler.
+func NewResourceHandler(
+	client dynamic.Interface,
+	bindingInfo *BindingInfo,
+	resource unstructured.Unstructured,
+	relatedGroupVersionResource schema.GroupVersionResource,
+	inputPathPrefix *string,
+	restMapper meta.RESTMapper,
+) (*ResourceHandler, error) {
+	if client == nil {
+		return nil, InvalidArgumentErr("client")
+	}
+
+	if bindingInfo == nil {
+		return nil, InvalidArgumentErr("bindingInfo")
+	}
+
+	if len(bindingInfo.SourcePath) == 0 {
+		return nil, InvalidArgumentErr("bindingInfo.Path")
+	}
+
+	if len(bindingInfo.ResourceReferencePath) == 0 {
+		return nil, InvalidArgumentErr("bindingInfo.ResourceReferencePath")
```

Comment:
Shouldn't these errors be `EmptyArgumentErr` instead of `InvalidArgumentErr`

Reply:
The intention is to return an error indicating that the argument is invalid, regardless of the validation error, pretty much like C#'s `ArgumentException` or Java's `IllegalArgumentException`.

Something along the lines of the following example:

```go
type ArgumentErr string

func (e ArgumentErr) Error() string {
    return fmt.Sprintf("argument error: %s" , string(e))
}
```

Usage would be:

```go
if len(someVar) > 0 {
    return ArgumentErr("should be greater than 0")
}
```

I haven't researched to see if there's some popular library that does that already.